<a href="https://colab.research.google.com/github/florianaewing/CSB430SWIWinter2026/blob/main/OptimumDeparture_by_Time_and_LocationPredictor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Florian Ewing, Margaret Keau, John Farnandez, Lauren Connely
# Professor Ix Procopios
# Software Design and Implementation
# 2.11.2026

#Optimum Departure Location & Airline by Region Predictor

This project analyzes large-scale U.S. flight data from 2023 and 2024 to build a machine learning model that predicts whether a flight will depart on time or experience a significant delay.

The model uses a binary classification target:

## Class 0 → Flight departed early or within 15 minutes of
##scheduled departure

##Class 1 → Flight departed 15 minutes or more after scheduled departure

Using this prediction capability, the system evaluates airport and airline performance for a given date and recommends the optimal airport and airline combination with the lowest probability of delay.

# Data Loading and Preprocessing
The dataset contains multiple CSV files that provide complementary information about US civil flights. The main file, US_flights_2023.csv, records over 6.7 million flights with details on departure and arrival times, delays, aircraft information, and distances.

Cancelled_Diverted_2023.csv focuses on flight cancellations and diversions, with just over 100,000 entries.

maj us flight - january 2024.csv contains approximately half a million flights from January 2024, including aircraft manufacturer and age.

airports_geolocation.csv provides latitude and longitude for 364 airports, along with city and state information.

Finally, weather_meteo_by_airport.csv contains daily weather measurements for airports, including temperature, precipitation, wind, and pressure.

In [1]:
import os
import pandas as pd
import numpy as np
import kagglehub

# ==== Load dataset ====
dataset_path = kagglehub.dataset_download(
    "bordanova/2023-us-civil-flights-delay-meteo-and-aircraft"
)

flights_2023 = pd.read_csv(os.path.join(dataset_path, "US_flights_2023.csv"))
flights_jan2024 = pd.read_csv(
    os.path.join(dataset_path, "maj us flight - january 2024.csv")
)
cancelled_diverted = pd.read_csv(
    os.path.join(dataset_path, "Cancelled_Diverted_2023.csv")
)
airports = pd.read_csv(os.path.join(dataset_path, "airports_geolocation.csv"))
weather = pd.read_csv(os.path.join(dataset_path, "weather_meteo_by_airport.csv"))

# ==== Fix types and typos ====
for df in [flights_2023, flights_jan2024, cancelled_diverted, weather]:
    if "FlightDate" in df.columns:
        df["FlightDate"] = pd.to_datetime(df["FlightDate"])
    if "time" in df.columns:
        df["time"] = pd.to_datetime(df["time"])

flights_2023.rename(columns={"Aicraft_age": "Aircraft_age"}, inplace=True)
flights_jan2024.rename(columns={"Aicraft_age": "Aircraft_age"}, inplace=True)

# ==== Combine flights ====
flights = pd.concat([flights_2023, flights_jan2024], ignore_index=True)

# ==== Merge airport geolocation ====
airports_geo = airports[["IATA_CODE", "LATITUDE", "LONGITUDE"]]

# Departure
flights = flights.merge(airports_geo, left_on="Dep_Airport",
                        right_on="IATA_CODE", how="left")
flights.rename(columns={"LATITUDE": "Dep_LAT", "LONGITUDE": "Dep_LON"},
               inplace=True)
flights.drop(columns=["IATA_CODE"], inplace=True)

# Arrival
flights = flights.merge(airports_geo, left_on="Arr_Airport",
                        right_on="IATA_CODE", how="left")
flights.rename(columns={"LATITUDE": "Arr_LAT", "LONGITUDE": "Arr_LON"},
               inplace=True)
flights.drop(columns=["IATA_CODE"], inplace=True)

# ==== Merge weather ====
def merge_weather(df, w_df, prefix, airport_col):
    w = w_df.copy()
    w.rename(columns={"time": "FlightDate", "airport_id": airport_col},
             inplace=True)
    rename_cols = {c: f"{prefix}_{c}" for c in w.columns
                   if c not in ["FlightDate", airport_col]}
    w.rename(columns=rename_cols, inplace=True)
    return df.merge(w, on=["FlightDate", airport_col], how="left")

flights = merge_weather(flights, weather, "DepWeather", "Dep_Airport")
flights = merge_weather(flights, weather, "ArrWeather", "Arr_Airport")

# ==== Merge cancellations ====
cancelled_diverted["FlightNumber"] = cancelled_diverted["Tail_Number"]
cancelled_diverted["Cancelled_Class"] = cancelled_diverted["Cancelled"].fillna(0)

flights = flights.merge(
    cancelled_diverted[['FlightDate','Airline','FlightNumber',
                        'Dep_Airport','Cancelled_Class']],
    left_on=['FlightDate','Airline','Tail_Number','Dep_Airport'],
    right_on=['FlightDate','Airline','FlightNumber','Dep_Airport'],
    how='left'
)
flights['Cancelled_Class'] = flights['Cancelled_Class'].fillna(0)

# ==== Create delay class ====
flights = flights.dropna(subset=["Dep_Delay"])
flights["Delay_Class"] = np.where(flights["Dep_Delay"] >= 15, 1, 0)

# ==== Summary stats ====
print("Delay Class Counts:")
print(flights["Delay_Class"].value_counts())
print("\nDelay Class %:")
print((flights["Delay_Class"].value_counts(normalize=True)*100).round(2))

# Cancellation summary
print("\nCancellation Counts:")
print(flights["Cancelled_Class"].value_counts())
print("\nCancellation %:")
print((flights["Cancelled_Class"].value_counts(normalize=True)*100).round(2))


100%|██████████| 127M/127M [00:01<00:00, 123MB/s]

Extracting files...


Delay Class Counts:
Delay_Class
0    5780639
1    1492770
Name: count, dtype: int64

Delay Class %:
Delay_Class
0    79.48
1    20.52
Name: proportion, dtype: float64

Cancellation Counts:
Cancelled_Class
0.0    7231140
1.0      42269
Name: count, dtype: int64

Cancellation %:
Cancelled_Class
0.0    99.42
1.0     0.58
Name: proportion, dtype: float64


# Key Metrics from each File

airports_geolocation.csv has 364 complete rows with location data and no duplicates.

Cancelled_Diverted_2023.csv has 104,488 rows, no missing values, 945 duplicates, mostly cancelled flights, skewed delay distributions, 15 airlines, and ~345 airports.

maj us flight - january 2024.csv contains 527,197 rows, no missing values, three duplicates, skewed delay metrics, and aircraft details (manufacturer, model, age ~14 years).

US_flights_2023.csv has 6,743,404 rows, no missing values, 31 duplicates, similar delay distributions, and consistent categorical data.

weather_meteo_by_airport.csv has 132,860 rows, no missing values, and complete weather metrics per airport

In the flight datasets (Cancelled_Diverted_2023.csv, maj us flight - january 2024.csv, US_flights_2023.csv), median departure and arrival delays are close to zero (or negative in January 2024), while maximum delays reach thousands of minutes (e.g., 3,024–4,413 minutes). This suggests that most flights have minimal delay, but a small number of flights experience extreme delays.

Delay components broken down by cause—
carrier,
weather,
NAS,
security,
last aircraft
—show similar patterns: mostly zero or very low values, with occasional extreme spikes.

# Data Processing and Visualization for Information Specific to Flight Delays

This code loads and merges U.S. civil flight, airport, and weather datasets from 2023 and January 2024, standardizes column names, and converts date fields to datetime objects. It adds latitude and longitude for departure and arrival airports and attaches corresponding weather data by date and airport. The code then calculates descriptive statistics and skewness for key delay metrics, computes median departure delays by airline and airport, and evaluates correlations between departure delays and weather variables. The result is a cleaned, merged dataset with initial analyses that summarize flight delays and their potential relationship to weather.

# US Flights 2023–Jan 2024: Delay Analysis

## Dataset
- 7,273,409 flights with 46 features including flight, weather, and
  airport data.

## Delay Statistics
- **Median departure delay:** ~0 minutes for most airlines;
  Southwest Airlines has 1 min.
- **Distribution:** Highly skewed with extreme delays (max ~4,400 min).
- **Airport hotspots:** PHF (10.5 min), SCK (9 min) have the highest
  median delays.

## Weather Impact
- Weak correlations with departure delays (highest: precipitation
  r≈0.064).
- Weather alone explains very little of the variation in delays.

## Conclusion
- Most flights depart on time or early.
- A few extreme delays drive high skewness in the data.
- Dataset is ready for feature selection and predictive modeling,
  including cancellation information.


In [2]:
import numpy as np

# ===== Define target variables =====

# Drop rows with missing departure delays
flights = flights.dropna(subset=["Dep_Delay"])

# Binary target for departure delay
# 0 -> On time / < 15 min delay
# 1 -> Delayed >= 15 min
flights["Delay_Class"] = np.where(
    flights["Dep_Delay"] >= 15, 1, 0
)

# Delay class distribution
delay_counts = flights["Delay_Class"].value_counts()
delay_percent = flights["Delay_Class"].value_counts(normalize=True) * 100

print("Delay Class Counts:")
print(delay_counts)
print("\nDelay Class Percentage:")
print(delay_percent.round(2))

# Compute class weights for imbalance
total = len(flights)
weight_0 = total / (2 * delay_counts[0])
weight_1 = total / (2 * delay_counts[1])

class_weights = {0: weight_0, 1: weight_1}
print("\nSuggested Delay Class Weights:")
print(class_weights)

# ===== Cancellation target summary =====
# Only if cancellations exist
if "Cancelled_Class" in flights.columns:
    cancel_counts = flights["Cancelled_Class"].value_counts()
    cancel_percent = flights["Cancelled_Class"].value_counts(
        normalize=True
    ) * 100

    print("\nCancellation Class Counts:")
    print(cancel_counts)
    print("\nCancellation Class Percentage:")
    print(cancel_percent.round(2))


Delay Class Counts:
Delay_Class
0    5780639
1    1492770
Name: count, dtype: int64

Delay Class Percentage:
Delay_Class
0    79.48
1    20.52
Name: proportion, dtype: float64

Suggested Delay Class Weights:
{0: np.float64(0.6291180784684877), 1: np.float64(2.4362122095165364)}

Cancellation Class Counts:
Cancelled_Class
0.0    7231140
1.0      42269
Name: count, dtype: int64

Cancellation Class Percentage:
Cancelled_Class
0.0    99.42
1.0     0.58
Name: proportion, dtype: float64


# Flight Delay and Cancellation Summary (2023–2024)

## Departure Delay

- **Class 0 (On time / < 15 min delay):** 5,780,639 flights (79.48%)
- **Class 1 (Delayed ≥ 15 min):** 1,492,770 flights (20.52%)

**Suggested class weights for modeling (to handle imbalance):**
- 0 → 0.63
- 1 → 2.44

> Most flights are on time, but a significant 1 in 5 flights is delayed.

## Cancellations

- **Class 0 (Not cancelled):** 7,231,140 flights (99.42%)
- **Class 1 (Cancelled):** 42,269 flights (0.58%)

> Cancellations are rare, making them highly imbalanced for modeling.



In [3]:
# ===== Cell 1: Region Mapping =====

region_to_states = {
    "Northeast": ["ME","NH","VT","MA","RI","CT","NY","NJ","PA","MD","DE"],
    "Southeast": ["VA","WV","KY","TN","NC","SC","GA","FL","AL","MS","AR","LA"],
    "Midwest":   ["OH","IN","IL","MI","WI","MN","IA","MO","ND","SD","NE","KS"],
    "Southwest": ["TX","OK","NM","AZ"],
    "West":      ["CO","WY","MT","ID","UT","NV","CA","OR","WA","AK","HI"],
}

# Invert to state -> region lookup
state_to_region = {
    state: region
    for region, states in region_to_states.items()
    for state in states
}

# Tag airports with region
airports["Region"] = airports["STATE"].map(state_to_region)
airports_region = airports[["IATA_CODE", "Region"]].dropna()

print(airports_region["Region"].value_counts())

Region
West         106
Midwest       87
Southeast     82
Northeast     41
Southwest     40
Name: count, dtype: int64


In Cell 2, raw flight dates are decomposed into recurring temporal features that the model can learn reliable patterns from. Rather than using the calendar date itself (which appears too infrequently to generalize), each flight is tagged with its **month** (capturing seasonal trends like winter storms or summer travel surges), **day of week** (capturing weekly rhythms like busy Fridays), a **weekend flag**, and a **holiday proximity flag** that marks flights within 2 days of a major U.S. holiday. These features repeat consistently across the full dataset, giving the model enough examples to learn meaningful delay patterns tied to time.

 Instead of using the raw calendar date (e.g. "January 1st"), which only appears once or twice in the dataset and isn't enough for the model to learn from, we break the date down into smaller, recurring components that appear thousands of times across the data:

Month — captures seasonal patterns, like more delays in winter due to storms or in summer due to high travel volume

Day of week — captures weekly rhythms, like Fridays and Mondays typically having more delays than Tuesdays

Weekend flag — a simple yes/no that helps the model distinguish busier leisure travel days

Holiday proximity flag — marks any flight within 2 days of a major U.S. holiday, when airports tend to be significantly more congested

This way, when a user inputs a future date, we convert it into these same components and the model can make a meaningful prediction based on patterns it has actually seen enough times to learn reliably.

In [11]:
# ===== Cell 2: Feature Engineering =====
import pandas as pd
from sklearn.preprocessing import LabelEncoder

# ---- Holiday proximity flag ----
us_holidays = pd.to_datetime(
    [
        "2023-01-01",
        "2023-01-16",
        "2023-02-20",
        "2023-05-29",
        "2023-07-04",
        "2023-09-04",
        "2023-11-23",
        "2023-12-25",
        "2024-01-01",
        "2024-01-15",
        "2024-02-19",
        "2024-05-27",
        "2024-07-04",
        "2024-09-02",
        "2024-11-28",
        "2024-12-25",
    ]
)

def is_near_holiday(date_series, holidays, window=2):
    """Flag dates within `window` days of a major holiday."""
    flags = np.zeros(len(date_series), dtype=int)
    for h in holidays:
        diff = (date_series - h).dt.days.abs()
        flags |= (diff <= window).astype(int).values
    return flags

# ---- Extract temporal features ----
flights["Month"] = flights["FlightDate"].dt.month
flights["DayOfWeek"] = flights["FlightDate"].dt.dayofweek  # 0=Mon, 6=Sun
flights["IsWeekend"] = (flights["DayOfWeek"] >= 5).astype(int)
flights["IsHoliday"] = is_near_holiday(flights["FlightDate"], us_holidays)

# ---- Merge region onto flights ----
# Create a mapping series from airports_region for efficiency
region_map = airports_region.set_index('IATA_CODE')['Region']
flights['Region'] = flights['Dep_Airport'].map(region_map)

# ---- Select features (no leakage) ----
FEATURE_COLS = [
    "Airline",
    "Dep_Airport",
    "Month",
    "DayOfWeek",
    "IsWeekend",
    "IsHoliday",
    "Aircraft_age",
    "Dep_LAT",
    "Dep_LON",
    "DepWeather_tavg",  # Changed from DepWeather_temperature_2m_mean
    "DepWeather_prcp",  # Changed from DepWeather_precipitation_sum
    "DepWeather_wspd",  # Changed from DepWeather_windspeed_10m_max
]

# Drop rows missing any feature
model_df = flights[FEATURE_COLS + ["Delay_Class", "Region"]].dropna()

# ---- Encode categoricals ----
le_airline = LabelEncoder()
le_airport = LabelEncoder()

model_df = model_df.copy()
model_df["Airline_enc"] = le_airline.fit_transform(model_df["Airline"])
model_df["Dep_Airport_enc"] = le_airport.fit_transform(
    model_df["Dep_Airport"]
)

FINAL_FEATURES = [
    "Airline_enc",
    "Dep_Airport_enc",
    "Month",
    "DayOfWeek",
    "IsWeekend",
    "IsHoliday",
    "Aircraft_age",
    "Dep_LAT",
    "Dep_LON",
    "DepWeather_tavg",  # Changed
    "DepWeather_prcp",  # Changed
    "DepWeather_wspd",  # Changed
]

X = model_df[FINAL_FEATURES]
y = model_df["Delay_Class"]

print(f"Feature matrix shape: {X.shape}")
print(f"Class distribution:\n{y.value_counts(normalize=True).round(3)}")

Feature matrix shape: (7226742, 12)
Class distribution:
Delay_Class
0    0.795
1    0.205
Name: proportion, dtype: float64


In [12]:
# ===== Cell 3: Train Random Forest =====
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# ---- Sample down for Colab feasibility ----
# Stratify to preserve class balance in the sample
sample_df = model_df.sample(n=1_500_000, random_state=42)
X_s = sample_df[FINAL_FEATURES]
y_s = sample_df["Delay_Class"]

X_train, X_test, y_train, y_test = train_test_split(
    X_s, y_s, test_size=0.2, random_state=42, stratify=y_s
)

# ---- Train ----
rf = RandomForestClassifier(
    n_estimators=200,
    max_depth=20,
    min_samples_leaf=50,
    class_weight={0: 0.63, 1: 2.44},
    n_jobs=-1,
    random_state=42,
)

rf.fit(X_train, y_train)

# ---- Evaluate ----
y_pred = rf.predict(X_test)
print(classification_report(y_test, y_pred, target_names=["On Time","Delayed"]))

              precision    recall  f1-score   support

     On Time       0.86      0.69      0.77    238488
     Delayed       0.32      0.58      0.42     61512

    accuracy                           0.67    300000
   macro avg       0.59      0.63      0.59    300000
weighted avg       0.75      0.67      0.69    300000



In [13]:
# ===== Cell 4: Recommender Function =====

def get_top5_recommendations(region, date_str):
    """
    Given a region name and a date string (YYYY-MM-DD),
    return the top 5 airport+airline combos with lowest delay probability.
    """
    date = pd.to_datetime(date_str)

    # Temporal features from the date
    month      = date.month
    dow        = date.dayofweek
    is_weekend = int(dow >= 5)
    is_holiday = int(any(abs((date - h).days) <= 2 for h in us_holidays))

    # Filter airports in the selected region
    regional_airports = airports_region[
        airports_region["Region"] == region
    ]["IATA_CODE"].tolist()

    if not regional_airports:
        print(f"No airports found for region: {region}")
        return None

    # Get all airport+airline combos that actually operated in this region
    combos = (
        model_df[model_df["Dep_Airport"].isin(regional_airports)]
        [["Dep_Airport", "Airline"]]
        .drop_duplicates()
        .reset_index(drop=True)
    )

    if combos.empty:
        print("No flight combinations found for this region.")
        return None

    # Build a representative feature row per combo
    # Use median weather/geo/aircraft values per airport from historical data
    airport_medians = (
      model_df[model_df["Dep_Airport"].isin(regional_airports)]
      .groupby("Dep_Airport")[[
          "Aircraft_age",
          "Dep_LAT", "Dep_LON",
          "DepWeather_tavg",
          "DepWeather_prcp",
          "DepWeather_wspd",
      ]].median()
      .reset_index()
  )

    combos = combos.merge(airport_medians, on="Dep_Airport", how="left")

    # Add temporal features
    combos["Month"]      = month
    combos["DayOfWeek"]  = dow
    combos["IsWeekend"]  = is_weekend
    combos["IsHoliday"]  = is_holiday

    # Encode using the fitted label encoders
    # Handle unseen labels gracefully
    def safe_encode(encoder, values):
        known = set(encoder.classes_)
        return np.array([
            encoder.transform([v])[0] if v in known else -1
            for v in values
        ])

    combos["Airline_enc"]     = safe_encode(le_airline,  combos["Airline"])
    combos["Dep_Airport_enc"] = safe_encode(le_airport, combos["Dep_Airport"])

    # Drop combos with unknown encodings
    combos = combos[
        (combos["Airline_enc"] != -1) &
        (combos["Dep_Airport_enc"] != -1)
    ]

    # Predict delay probability for each combo
    X_pred = combos[FINAL_FEATURES].values
    delay_probs = rf.predict_proba(X_pred)[:, 1]  # P(delayed)

    combos["Delay_Probability"] = delay_probs
    combos["OnTime_Probability"] = 1 - delay_probs

    # Return top 5 sorted by lowest delay probability
    top5 = (
        combos[["Dep_Airport", "Airline", "Delay_Probability", "OnTime_Probability"]]
        .sort_values("Delay_Probability")
        .head(5)
        .reset_index(drop=True)
    )
    top5.index += 1  # rank from 1
    top5["Delay_Probability"]  = top5["Delay_Probability"].map("{:.1%}".format)
    top5["OnTime_Probability"] = top5["OnTime_Probability"].map("{:.1%}".format)

    return top5

In [15]:
# ===== Cell 5: User Interaction =====

VALID_REGIONS = list(region_to_states.keys())

print("=== Flight Delay Predictor ===")
print(f"Available regions: {', '.join(VALID_REGIONS)}\n")

region_input = input("Enter your departure region: ").strip().title()
date_input   = input("Enter your departure date (YYYY-MM-DD): ").strip()

if region_input not in VALID_REGIONS:
    print(f"Invalid region. Please choose from: {', '.join(VALID_REGIONS)}")
else:
    try:
        results = get_top5_recommendations(region_input, date_input)
        if results is not None:
            print(f"\nTop 5 airport+airline combinations for the {region_input} on {date_input}:\n")
            print(results.to_string())
    except Exception as e:
        print(f"Something went wrong: {e}")

=== Flight Delay Predictor ===
Available regions: Northeast, Southeast, Midwest, Southwest, West

Enter your departure region: West
Enter your departure date (YYYY-MM-DD): 2026-03-15

Top 5 airport+airline combinations for the West on 2026-03-15:

  Dep_Airport                Airline Delay_Probability OnTime_Probability
1         HLN  Skywest Airlines Inc.             22.6%              77.4%
2         GTF  Skywest Airlines Inc.             23.4%              76.6%
3         LWS  Skywest Airlines Inc.             23.8%              76.2%
4         BOI  Skywest Airlines Inc.             25.8%              74.2%
5         SEA  Skywest Airlines Inc.             26.2%              73.8%


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
